src/py21cmfast/src/ps.c

```diff
@@ -1555,35 +1568,122 @@ int ComputeLF(int nbins, struct UserParams *user_params, struct CosmoParams *cos
+        	}
+		}
+        else { 
+        	lnM_temp = calloc(nbins_smth,sizeof(double));
+        	deriv_temp = calloc(nbins_smth,sizeof(double));
+        	deriv = calloc(nbins,sizeof(double));
+
+        	for (i=0; i<nbins; i++) {
+            	// calculate luminosity function 
+            	lnMhalo_i = lnMhalo_lo + dlnM*(double)i;
+            	Mhalo_param[i] = exp(lnMhalo_i);
+
+            	M_h_z[i + i_z*nbins] = Mhalo_param[i];
+
+            	Muv_1 = gsl_spline_eval(LF_spline, lnMhalo_i - delta_lnMhalo, LF_spline_acc);
+            	Muv_2 = gsl_spline_eval(LF_spline, lnMhalo_i + delta_lnMhalo, LF_spline_acc);
+
+            	dMuvdMhalo = (Muv_2 - Muv_1) / (2.*delta_lnMhalo * exp(lnMhalo_i));
+            	deriv[i] = fabs(dMuvdMhalo);
+        	}
+
+        	deriv_spline_acc = gsl_interp_accel_alloc();
+        	deriv_spline = gsl_spline_alloc(gsl_interp_cspline, nbins_smth);
+        	//deriv_spline = gsl_spline_alloc(gsl_interp_steffen, nbins_smth);
+
+        	// generate interpolation arrays to smooth discontinuity of the derivative causing a kink
+        	// Note that the number of array elements and the range of interpolation are made by arbitrary choices.
+        	lnM_temp[0] = lnMhalo_param[i_unity - 3];
+        	lnM_temp[1] = lnMhalo_param[i_unity - 2];
+        	lnM_temp[2] = lnMhalo_param[i_unity + 8];
+        	lnM_temp[3] = lnMhalo_param[i_unity + 9];
+        	lnM_temp[4] = lnMhalo_param[i_unity + 10];
+        	lnM_temp[5] = lnMhalo_param[i_unity + 11];
+        	lnM_temp[6] = lnMhalo_param[i_unity + 12];
+
+        	deriv_temp[0] = deriv[i_unity - 3];
+        	deriv_temp[1] = deriv[i_unity - 2];
+        	deriv_temp[2] = deriv[i_unity + 8];
+        	deriv_temp[3] = deriv[i_unity + 9];
+        	deriv_temp[4] = deriv[i_unity + 10];
+        	deriv_temp[5] = deriv[i_unity + 11];
+        	deriv_temp[6] = deriv[i_unity + 12];
+
+        	gsl_spline_init(deriv_spline, lnM_temp, deriv_temp, nbins_smth);
+
+        	deriv[i_unity - 1] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity - 1], deriv_spline_acc);
+        	deriv[i_unity] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity], deriv_spline_acc);
+        	deriv[i_unity + 1] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 1], deriv_spline_acc);
+        	deriv[i_unity + 2] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 2], deriv_spline_acc);
+        	deriv[i_unity + 3] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 3], deriv_spline_acc);
+        	deriv[i_unity + 4] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 4], deriv_spline_acc);
+        	deriv[i_unity + 5] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 5], deriv_spline_acc);
+        	deriv[i_unity + 6] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 6], deriv_spline_acc);
+        	deriv[i_unity + 7] = gsl_spline_eval(deriv_spline, lnMhalo_param[i_unity + 7], deriv_spline_acc);
+
+        	for (i=0; i<nbins; i++) {
```

Comment:
I think we could even move this whole set of cases outside the if statement, and just divide by the deriv on the `if`.

Reply:
I think it is better to leave this part here. When it doesn't smooth LFs, the code doesn't generate 'deriv' array (using directly 'dMuvdMhalo'). So, if we move this part outside the if statement, we should generate 'deriv' array when the interpolation is not necessary.